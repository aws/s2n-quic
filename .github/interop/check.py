#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
#  SPDX-License-Identifier: Apache-2.0

import argparse
import json
import itertools
import sys

parser = argparse.ArgumentParser(description='Check interop reports.')
parser.add_argument('--required', type=str)
parser.add_argument('--regressions')
parser.add_argument('report')
args = parser.parse_args()

status = {'ok': True, 'failures': 0}
def fail(message):
    print(message)
    status['ok'] = False
    status['failures'] += 1

def format_required_report(report):
    outcome = {}
    result_idx = 0

    # find the version of s2n-quic to check
    s2n_quic = 's2n-quic'
    for impl in itertools.chain(report['clients'], report['servers']):
        # if we're testing a PR then use that name
        if impl.startswith('s2n-quic-pr'):
            s2n_quic = impl
            break

    for client in report['clients']:
        for server in report['servers']:
            result = report['results'][result_idx]
            result_idx += 1

            # we're only interested in s2n-quic results
            if client != s2n_quic and server != s2n_quic:
                continue

            for test in result:
                outcome.setdefault(test['name'], {})

                info = outcome[test['name']]
                info.setdefault(client, {'client': False, 'server': False})
                info.setdefault(server, {'client': False, 'server': False})

                success = test['result'] == 'succeeded'
                info[client]['client'] = success
                info[server]['server'] = success

    return outcome

with open(args.report) as f:
    result = json.load(f)

    if args.regressions and result['regression']:
        fail("A regression from main was detected")

    if args.required:
        with open(args.required) as r:
            required = json.load(r)

            actual = format_required_report(result)

            for test, impls in required.items():
                test_results = actual[test]
                for impl_name, endpoints in impls.items():
                    impl = test_results[impl_name]
                    for endpoint in endpoints:
                        if not impl[endpoint]:
                            fail("{} ({}) - {} was expected to pass but failed".format(impl_name, endpoint, test))

if not status['ok']:
    sys.exit(status['failures'])
