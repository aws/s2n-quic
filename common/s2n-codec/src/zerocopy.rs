// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use ::byteorder::NetworkEndian;
use core::{
    cmp::Ordering,
    fmt,
    hash::{Hash, Hasher},
};
pub use zerocopy::*;

#[cfg(feature = "generator")]
use bolero_generator::*;

/// Define a codec implementation for a zerocopy value that implements
/// `FromBytes`, `AsBytes`, and `Unaligned`.
#[macro_export]
macro_rules! zerocopy_value_codec {
    ($name:ident) => {
        impl<'a> $crate::DecoderValue<'a> for $name {
            fn decode(buffer: $crate::DecoderBuffer<'a>) -> $crate::DecoderBufferResult<Self> {
                let (value, buffer) = <&'a $name as $crate::DecoderValue>::decode(buffer)?;
                Ok((*value, buffer))
            }
        }

        impl<'a> $crate::DecoderValue<'a> for &'a $name {
            fn decode(buffer: $crate::DecoderBuffer<'a>) -> $crate::DecoderBufferResult<Self> {
                let (value, buffer) =
                    $crate::zerocopy::LayoutVerified::<_, $name>::new_unaligned_from_prefix(
                        buffer.into_less_safe_slice(),
                    )
                    .ok_or_else(|| $crate::DecoderError::InvariantViolation("invalid payload"))?;

                Ok((value.into_ref(), buffer.into()))
            }
        }

        impl<'a> $crate::DecoderValueMut<'a> for $name {
            fn decode_mut(
                buffer: $crate::DecoderBufferMut<'a>,
            ) -> $crate::DecoderBufferMutResult<Self> {
                let (value, buffer) = <&'a $name as $crate::DecoderValueMut>::decode_mut(buffer)?;
                Ok((*value, buffer))
            }
        }

        impl<'a> $crate::DecoderValueMut<'a> for &'a $name {
            fn decode_mut(
                buffer: $crate::DecoderBufferMut<'a>,
            ) -> $crate::DecoderBufferMutResult<'a, Self> {
                let (value, buffer) =
                    <&'a mut $name as $crate::DecoderValueMut>::decode_mut(buffer)?;
                Ok((value, buffer))
            }
        }

        impl<'a> $crate::DecoderValueMut<'a> for &'a mut $name {
            fn decode_mut(
                buffer: $crate::DecoderBufferMut<'a>,
            ) -> $crate::DecoderBufferMutResult<'a, Self> {
                let (value, buffer) =
                    $crate::zerocopy::LayoutVerified::<_, $name>::new_unaligned_from_prefix(
                        buffer.into_less_safe_slice(),
                    )
                    .ok_or_else(|| $crate::DecoderError::InvariantViolation("invalid payload"))?;

                Ok((value.into_mut(), buffer.into()))
            }
        }

        impl $crate::EncoderValue for $name {
            fn encoding_size(&self) -> usize {
                core::mem::size_of::<$name>()
            }

            fn encoding_size_for_encoder<E: $crate::Encoder>(&self, _encoder: &E) -> usize {
                core::mem::size_of::<$name>()
            }

            fn encode<E: $crate::Encoder>(&self, encoder: &mut E) {
                encoder.write_slice($crate::zerocopy::AsBytes::as_bytes(self));
            }
        }

        impl<'a> $crate::EncoderValue for &'a $name {
            fn encoding_size(&self) -> usize {
                core::mem::size_of::<$name>()
            }

            fn encoding_size_for_encoder<E: $crate::Encoder>(&self, _encoder: &E) -> usize {
                ::core::mem::size_of::<$name>()
            }

            fn encode<E: $crate::Encoder>(&self, encoder: &mut E) {
                encoder.write_slice($crate::zerocopy::AsBytes::as_bytes(*self));
            }
        }

        impl<'a> $crate::EncoderValue for &'a mut $name {
            fn encoding_size(&self) -> usize {
                core::mem::size_of::<$name>()
            }

            fn encoding_size_for_encoder<E: $crate::Encoder>(&self, _encoder: &E) -> usize {
                ::core::mem::size_of::<$name>()
            }

            fn encode<E: $crate::Encoder>(&self, encoder: &mut E) {
                encoder.write_slice($crate::zerocopy::AsBytes::as_bytes(*self));
            }
        }
    };
}

// The `zerocopy` crate provides integer types that are able to be referenced
// in an endian-independent method. This macro wraps those types and implements
// a few convenience traits.
macro_rules! zerocopy_network_integer {
    ($native:ident, $name:ident) => {
        #[derive(
            Clone,
            Copy,
            Default,
            Eq,
            $crate::zerocopy::FromBytes,
            $crate::zerocopy::AsBytes,
            $crate::zerocopy::Unaligned,
        )]
        #[repr(C)]
        pub struct $name(::zerocopy::byteorder::$name<NetworkEndian>);

        impl $name {
            pub const ZERO: Self = Self(::zerocopy::byteorder::$name::ZERO);

            #[inline]
            pub fn new(value: $native) -> Self {
                value.into()
            }

            #[inline]
            pub fn get(&self) -> $native {
                self.0.get()
            }

            #[inline]
            pub fn set(&mut self, value: $native) {
                self.0.set(value);
            }
        }

        impl PartialEq for $name {
            fn eq(&self, other: &Self) -> bool {
                self.cmp(other) == Ordering::Equal
            }
        }

        impl PartialEq<$native> for $name {
            fn eq(&self, other: &$native) -> bool {
                self.partial_cmp(other) == Some(Ordering::Equal)
            }
        }

        impl PartialOrd for $name {
            fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
                Some(self.cmp(other))
            }
        }

        impl PartialOrd<$native> for $name {
            fn partial_cmp(&self, other: &$native) -> Option<Ordering> {
                Some(self.0.get().cmp(other))
            }
        }

        impl Ord for $name {
            fn cmp(&self, other: &Self) -> Ordering {
                self.0.get().cmp(&other.0.get())
            }
        }

        impl Hash for $name {
            fn hash<H: Hasher>(&self, state: &mut H) {
                self.0.get().hash(state);
            }
        }

        impl fmt::Debug for $name {
            fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                write!(formatter, "{}", self.0.get())
            }
        }

        impl fmt::Display for $name {
            fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                write!(formatter, "{}", self.0.get())
            }
        }

        impl From<$native> for $name {
            fn from(value: $native) -> Self {
                Self(::zerocopy::byteorder::$name::new(value))
            }
        }

        impl From<$name> for $native {
            fn from(v: $name) -> $native {
                v.0.get()
            }
        }

        #[cfg(feature = "generator")]
        impl TypeGenerator for $name {
            fn generate<D: Driver>(driver: &mut D) -> Option<Self> {
                Some(Self::new(driver.gen()?))
            }
        }

        zerocopy_value_codec!($name);
    };
}

zerocopy_network_integer!(i16, I16);
zerocopy_network_integer!(u16, U16);
zerocopy_network_integer!(i32, I32);
zerocopy_network_integer!(u32, U32);
zerocopy_network_integer!(i64, I64);
zerocopy_network_integer!(u64, U64);
zerocopy_network_integer!(i128, I128);
zerocopy_network_integer!(u128, U128);

#[test]
fn zerocopy_struct_test() {
    use crate::DecoderBuffer;

    #[derive(Copy, Clone, Debug, PartialEq, PartialOrd, FromBytes, AsBytes, Unaligned)]
    #[repr(C)]
    struct UdpHeader {
        source_port: U16,
        destination_port: U16,
        payload_len: U16,
        checksum: U16,
    }

    zerocopy_value_codec!(UdpHeader);

    let buffer = vec![0, 1, 0, 2, 0, 3, 0, 4];
    let decoder = DecoderBuffer::new(&buffer);
    let (mut header, _) = decoder.decode().unwrap();

    ensure_codec_round_trip_value!(UdpHeader, header).unwrap();
    ensure_codec_round_trip_value!(&UdpHeader, &header).unwrap();
    ensure_codec_round_trip_value_mut!(&mut UdpHeader, &mut header).unwrap();

    assert_eq!(header.source_port, 1u16);
    assert_eq!(header.destination_port, 2u16);
    assert_eq!(header.payload_len, 3u16);
    assert_eq!(header.checksum, 4u16);
}
