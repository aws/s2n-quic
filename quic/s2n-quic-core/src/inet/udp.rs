// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use core::fmt;
use s2n_codec::zerocopy::U16;

// NOTE: duvet doesn't know how to parse this RFC since it doesn't follow more modern formatting
//# https://www.rfc-editor.org/rfc/rfc768#format
//# Format
//# ------
//#                  0      7 8     15 16    23 24    31
//#                 +--------+--------+--------+--------+
//#                 |     Source      |   Destination   |
//#                 |      Port       |      Port       |
//#                 +--------+--------+--------+--------+
//#                 |                 |                 |
//#                 |     Length      |    Checksum     |
//#                 +--------+--------+--------+--------+
//#                 |
//#                 |          data octets ...
//#                 +---------------- ...

define_inet_type!(
    pub struct Header {
        source: U16,
        destination: U16,
        len: U16,
        checksum: U16,
    }
);

impl Header {
    /// Swaps the direction of the header
    #[inline]
    pub fn swap(&mut self) {
        core::mem::swap(&mut self.source, &mut self.destination)
    }

    #[inline]
    pub const fn source(&self) -> &U16 {
        &self.source
    }

    #[inline]
    pub fn source_mut(&mut self) -> &mut U16 {
        &mut self.source
    }

    #[inline]
    pub const fn destination(&self) -> &U16 {
        &self.destination
    }

    #[inline]
    pub fn destination_mut(&mut self) -> &mut U16 {
        &mut self.destination
    }

    #[inline]
    pub const fn len(&self) -> &U16 {
        &self.len
    }

    #[inline]
    pub fn len_mut(&mut self) -> &mut U16 {
        &mut self.len
    }

    #[inline]
    pub const fn checksum(&self) -> &U16 {
        &self.checksum
    }

    #[inline]
    pub fn checksum_mut(&mut self) -> &mut U16 {
        &mut self.checksum
    }
}

impl fmt::Debug for Header {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_struct("udp::Header")
            .field("source", &self.source)
            .field("destination", &self.destination)
            .field("len", &self.len)
            .field("checksum", &format_args!("0x{:04x}", &self.checksum.get()))
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bolero::check;
    use s2n_codec::DecoderBuffer;

    #[test]
    #[cfg_attr(miri, ignore)]
    fn snapshot_test() {
        let mut buffer = vec![0u8; core::mem::size_of::<Header>()];
        for (idx, byte) in buffer.iter_mut().enumerate() {
            *byte = idx as u8;
        }
        let decoder = DecoderBuffer::new(&buffer);
        let (header, _) = decoder.decode::<&Header>().unwrap();
        insta::assert_debug_snapshot!("snapshot_test", header);

        for byte in &mut buffer {
            *byte = 255;
        }
        let decoder = DecoderBuffer::new(&buffer);
        let (header, _) = decoder.decode::<&Header>().unwrap();
        insta::assert_debug_snapshot!("snapshot_filled_test", header);
    }

    #[test]
    fn header_round_trip_test() {
        check!().for_each(|buffer| {
            s2n_codec::assert_codec_round_trip_bytes!(Header, buffer);
        });
    }
}
