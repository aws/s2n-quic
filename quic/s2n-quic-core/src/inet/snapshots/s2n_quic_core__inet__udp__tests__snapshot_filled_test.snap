---
source: quic/s2n-quic-core/src/inet/udp.rs
expression: header
---
udp::Header {
    source: 65535,
    destination: 65535,
    len: 65535,
    checksum: 0xffff,
}
